toolium/utils/dataset.py

```diff
@@ -236,18 +236,24 @@ def _replace_param_fixed_length(param):
 
 def _infer_param_type(param):
     """
-    Transform the param from string to the inferred data type
-    E.g. '1234' -> 1234, '0.50' -> 0.5, ["a", "b"]' -> ["a", "b"], '{"a": "b"}' -> {"a": "b"}
+    Transform the param from a string representing a built-in data type or a JSON object/list to the
+    corresponding built-in data type. If no conversion is possible, the original param is returned.
+    E.g. "1234" -> 1234, "0.50" -> 0.5, "True" -> True, "None" -> None,
+    "['a', None]" -> ['a', None], "{'a': None}" -> {'a': None},
+    '["a", null]' -> ["a", None], '{"a": null}' -> {"a": None}
 
     :param param: data to be transformed
     :return data with the inferred type
     """
     new_param = param
+    # attempt direct conversion to a built-in data type
     try:
         new_param = literal_eval(param)
     except Exception:
-        try:
-            new_param = json.loads(param)
-        except Exception:
-            pass
+        if param.startswith('{') or param.startswith('['):
```

Comment:
include 'if' line inside try block, it will fail if param is not a string

Reply:
Currently it is an internal function that is only called from replace_param when param is a string. The docstring also claims that param is expected to be a string. But alright, that may eventually change in the future... I will change it to be on the safe side.